rules/aip0123/resource_pattern.go

```diff
@@ -0,0 +1,68 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package aip0123
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/stoewer/go-strcase"
+
+	"github.com/googleapis/api-linter/locations"
+
+	"github.com/googleapis/api-linter/lint"
+	"github.com/googleapis/api-linter/rules/internal/utils"
+	"github.com/jhump/protoreflect/desc"
+)
+
+var resourcePattern = &lint.MessageRule{
+	Name:   lint.NewRuleName(123, "resource-pattern"),
+	OnlyIf: hasResourceAnnotation,
+	LintMessage: func(m *desc.MessageDescriptor) []lint.Problem {
+		resource := utils.GetResource(m)
+		// Are any patterns declared at all? If not, complain.
+		if len(resource.GetPattern()) == 0 {
+			return []lint.Problem{{
+				Message:    "Resources should declare resource name pattern(s).",
+				Descriptor: m,
+				Location:   locations.MessageResource(m),
+			}}
+		}
+
+		// Ensure that the constant segments of the pattern uses camel case,
+		// not snake case.
+		for _, pattern := range resource.GetPattern() {
+			if strings.Contains(getPlainPattern(pattern), "_") {
+				want := []string{}
+				for _, token := range strings.Split(pattern, "/") {
+					if strings.HasPrefix(token, "{") && strings.HasSuffix(token, "}") {
+						want = append(want, token)
+					} else {
+						want = append(want, strcase.LowerCamelCase(token))
+					}
+				}
```

Comment:
Ah, I did expect this complex and difficulty -- sorry! I think we should go back to your original version -- it is better for no suggestion than an incorrect suggestion, and I think, it is no worth to go extra miles now. We can wait until we can give a specific location. Thanks!

Reply:
did not expect

Reply:
Actually, I changed my mind in your favor. I refactored this out into a common method that all three spots can use, that should always give the correct pattern as a suggestion.

Reply:
Nice. Thank you!